<a href="https://colab.research.google.com/github/SANYA-95/summer-of-code-2023/blob/main/Feature_Engineering_and_final_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv('USA_cars_datasets-2.csv')
df

Unnamed: 0  price      brand    model  year   title_status   mileage  \
0              0   6300     toyota  cruiser  2008  clean vehicle  274117.0   
1              1   2899       ford       se  2011  clean vehicle  190552.0   
2              2   5350      dodge      mpv  2018  clean vehicle   39590.0   
3              3  25000       ford     door  2014  clean vehicle   64146.0   
4              4  27700  chevrolet     1500  2018  clean vehicle    6654.0   
...          ...    ...        ...      ...   ...            ...       ...   
2494        2494   7800     nissan    versa  2019  clean vehicle   23609.0   
2495        2495   9200     nissan    versa  2018  clean vehicle   34553.0   
2496        2496   9200     nissan    versa  2018  clean vehicle   31594.0   
2497        2497   9200     nissan    versa  2018  clean vehicle   32557.0   
2498        2498   9200     nissan    versa  2018  clean vehicle   31371.0   

       color                  vin        lot       state country  \
0      black    jtezu11f88k007763  159348797  new jersey     usa   
1     silver    2fmdk3gc4bbb02217  166951262   tennessee     usa   
2     silver    3c4pdcgg5jt346413  167655728     georgia     usa   
3       blue    1ftfw1et4efc23745  167753855    virginia     usa   
4        red    3gcpcrec2jg473991  167763266     florida     usa   
...      ...                  ...        ...         ...     ...   
2494     red    3n1cn7ap9kl880319  167722715  california     usa   
2495  silver    3n1cn7ap5jl884088  167762225     florida     usa   
2496  silver    3n1cn7ap9jl884191  167762226     florida     usa   
2497   black    3n1cn7ap3jl883263  167762227     florida     usa   
2498  silver    3n1cn7ap4jl884311  167762228     florida     usa   

          condition  
0      10 days left  
1       6 days left  
2       2 days left  
3     22 hours left  
4     22 hours left  
...             ...  
2494    1 days left  
2495  21 hours left  
2496  21 hours left  
2497    2 days left  
2498  21 hours left  

[2499 rows x 13 columns]

## **Data Cleaning**

As concluded from the last colab https://colab.research.google.com/drive/1rMoplAtMiNwbjCPiuGT5IsNo578lXNVN#scrollTo=TrdPbt9cfP8Y

I will drop all the columns except those on which price was dependent

In [ ]:
df = df.drop(['Unnamed: 0','mileage','vin','lot','condition'], axis='columns')
df.head()

price      brand    model  year   title_status   color       state country
0   6300     toyota  cruiser  2008  clean vehicle   black  new jersey     usa
1   2899       ford       se  2011  clean vehicle  silver   tennessee     usa
2   5350      dodge      mpv  2018  clean vehicle  silver     georgia     usa
3  25000       ford     door  2014  clean vehicle    blue    virginia     usa
4  27700  chevrolet     1500  2018  clean vehicle     red     florida     usa

# **Feature Engineering**

Since the machine model is comfortable in having the independent variables as numeric values, we need to change the variables such as brand, model, title_status, color, state and country in numeric values.

This will be done using ***One-Hot encoding*** and ***Dummy Variables***

### **DUMMY VARIABLES**

In [5]:
dummy1  = pd.get_dummies(df.brand)
dummy1

acura  audi  bmw  buick  cadillac  chevrolet  chrysler  dodge  ford  \
0         0     0    0      0         0          0         0      0     0   
1         0     0    0      0         0          0         0      0     1   
2         0     0    0      0         0          0         0      1     0   
3         0     0    0      0         0          0         0      0     1   
4         0     0    0      0         0          1         0      0     0   
...     ...   ...  ...    ...       ...        ...       ...    ...   ...   
2494      0     0    0      0         0          0         0      0     0   
2495      0     0    0      0         0          0         0      0     0   
2496      0     0    0      0         0          0         0      0     0   
2497      0     0    0      0         0          0         0      0     0   
2498      0     0    0      0         0          0         0      0     0   

      gmc  ...  land  lexus  lincoln  maserati  mazda  mercedes-benz  nissan  \
0       0  ...     0      0        0         0      0              0       0   
1       0  ...     0      0        0         0      0              0       0   
2       0  ...     0      0        0         0      0              0       0   
3       0  ...     0      0        0         0      0              0       0   
4       0  ...     0      0        0         0      0              0       0   
...   ...  ...   ...    ...      ...       ...    ...            ...     ...   
2494    0  ...     0      0        0         0      0              0       1   
2495    0  ...     0      0        0         0      0              0       1   
2496    0  ...     0      0        0         0      0              0       1   
2497    0  ...     0      0        0         0      0              0       1   
2498    0  ...     0      0        0         0      0              0       1   

      peterbilt  ram  toyota  
0             0    0       1  
1             0    0       0  
2             0    0       0  
3             0    0       0  
4             0    0       0  
...         ...  ...     ...  
2494          0    0       0  
2495          0    0       0  
2496          0    0       0  
2497          0    0       0  
2498          0    0       0  

[2499 rows x 28 columns]

In [6]:
dummy2 = pd.get_dummies(df.model)
dummy2

1500  2500  2500hd  300  3500  5  acadia  altima  armada  bus  ...  van  \
0        0     0       0    0     0  0       0       0       0    0  ...    0   
1        0     0       0    0     0  0       0       0       0    0  ...    0   
2        0     0       0    0     0  0       0       0       0    0  ...    0   
3        0     0       0    0     0  0       0       0       0    0  ...    0   
4        1     0       0    0     0  0       0       0       0    0  ...    0   
...    ...   ...     ...  ...   ... ..     ...     ...     ...  ...  ...  ...   
2494     0     0       0    0     0  0       0       0       0    0  ...    0   
2495     0     0       0    0     0  0       0       0       0    0  ...    0   
2496     0     0       0    0     0  0       0       0       0    0  ...    0   
2497     0     0       0    0     0  0       0       0       0    0  ...    0   
2498     0     0       0    0     0  0       0       0       0    0  ...    0   

      vans  vehicl  versa  volt  wagon  x3  xd  xt5  xterra  
0        0       0      0     0      0   0   0    0       0  
1        0       0      0     0      0   0   0    0       0  
2        0       0      0     0      0   0   0    0       0  
3        0       0      0     0      0   0   0    0       0  
4        0       0      0     0      0   0   0    0       0  
...    ...     ...    ...   ...    ...  ..  ..  ...     ...  
2494     0       0      1     0      0   0   0    0       0  
2495     0       0      1     0      0   0   0    0       0  
2496     0       0      1     0      0   0   0    0       0  
2497     0       0      1     0      0   0   0    0       0  
2498     0       0      1     0      0   0   0    0       0  

[2499 rows x 127 columns]

In [7]:
dummy3 = pd.get_dummies(df.title_status)
dummy3

clean vehicle  salvage insurance
0                 1                  0
1                 1                  0
2                 1                  0
3                 1                  0
4                 1                  0
...             ...                ...
2494              1                  0
2495              1                  0
2496              1                  0
2497              1                  0
2498              1                  0

[2499 rows x 2 columns]

In [8]:
dummy4= pd.get_dummies(df.color)
dummy4

beige  billet silver metallic clearcoat  black  black clearcoat  blue  \
0         0                                 0      1                0     0   
1         0                                 0      0                0     0   
2         0                                 0      0                0     0   
3         0                                 0      0                0     1   
4         0                                 0      0                0     0   
...     ...                               ...    ...              ...   ...   
2494      0                                 0      0                0     0   
2495      0                                 0      0                0     0   
2496      0                                 0      0                0     0   
2497      0                                 0      1                0     0   
2498      0                                 0      0                0     0   

      bright white clearcoat  brown  burgundy  cayenne red  charcoal  ...  \
0                          0      0         0            0         0  ...   
1                          0      0         0            0         0  ...   
2                          0      0         0            0         0  ...   
3                          0      0         0            0         0  ...   
4                          0      0         0            0         0  ...   
...                      ...    ...       ...          ...       ...  ...   
2494                       0      0         0            0         0  ...   
2495                       0      0         0            0         0  ...   
2496                       0      0         0            0         0  ...   
2497                       0      0         0            0         0  ...   
2498                       0      0         0            0         0  ...   

      silver  super black  tan  toreador red  triple yellow tri-coat  \
0          0            0    0             0                       0   
1          1            0    0             0                       0   
2          1            0    0             0                       0   
3          0            0    0             0                       0   
4          0            0    0             0                       0   
...      ...          ...  ...           ...                     ...   
2494       0            0    0             0                       0   
2495       1            0    0             0                       0   
2496       1            0    0             0                       0   
2497       0            0    0             0                       0   
2498       1            0    0             0                       0   

      turquoise  tuxedo black metallic  white  \
0             0                      0      0   
1             0                      0      0   
2             0                      0      0   
3             0                      0      0   
4             0                      0      0   
...         ...                    ...    ...   
2494          0                      0      0   
2495          0                      0      0   
2496          0                      0      0   
2497          0                      0      0   
2498          0                      0      0   

      white platinum tri-coat metallic  yellow  
0                                    0       0  
1                                    0       0  
2                                    0       0  
3                                    0       0  
4                                    0       0  
...                                ...     ...  
2494                                 0       0  
2495                                 0       0  
2496                                 0       0  
2497                                 0       0  
2498                                 0       0  

[2499 rows x 49 columns]

In [9]:
dummy5 = pd.get_dummies(df.state)
dummy5

alabama  arizona  arkansas  california  colorado  connecticut  florida  \
0           0        0         0           0         0            0        0   
1           0        0         0           0         0            0        0   
2           0        0         0           0         0            0        0   
3           0        0         0           0         0            0        0   
4           0        0         0           0         0            0        1   
...       ...      ...       ...         ...       ...          ...      ...   
2494        0        0         0           1         0            0        0   
2495        0        0         0           0         0            0        1   
2496        0        0         0           0         0            0        1   
2497        0        0         0           0         0            0        1   
2498        0        0         0           0         0            0        1   

      georgia  idaho  illinois  ...  south carolina  tennessee  texas  utah  \
0           0      0         0  ...               0          0      0     0   
1           0      0         0  ...               0          1      0     0   
2           1      0         0  ...               0          0      0     0   
3           0      0         0  ...               0          0      0     0   
4           0      0         0  ...               0          0      0     0   
...       ...    ...       ...  ...             ...        ...    ...   ...   
2494        0      0         0  ...               0          0      0     0   
2495        0      0         0  ...               0          0      0     0   
2496        0      0         0  ...               0          0      0     0   
2497        0      0         0  ...               0          0      0     0   
2498        0      0         0  ...               0          0      0     0   

      vermont  virginia  washington  west virginia  wisconsin  wyoming  
0           0         0           0              0          0        0  
1           0         0           0              0          0        0  
2           0         0           0              0          0        0  
3           0         1           0              0          0        0  
4           0         0           0              0          0        0  
...       ...       ...         ...            ...        ...      ...  
2494        0         0           0              0          0        0  
2495        0         0           0              0          0        0  
2496        0         0           0              0          0        0  
2497        0         0           0              0          0        0  
2498        0         0           0              0          0        0  

[2499 rows x 44 columns]

In [10]:
dummy6 = pd.get_dummies(df.country)
dummy6

canada   usa
0           0     1
1           0     1
2           0     1
3           0     1
4           0     1
...       ...   ...
2494        0     1
2495        0     1
2496        0     1
2497        0     1
2498        0     1

[2499 rows x 2 columns]

In [12]:
merged = pd.concat([df,dummy1, dummy2,dummy3,dummy4,dummy5,dummy6], axis='columns')
merged

price      brand    model  year   title_status   color       state  \
0      6300     toyota  cruiser  2008  clean vehicle   black  new jersey   
1      2899       ford       se  2011  clean vehicle  silver   tennessee   
2      5350      dodge      mpv  2018  clean vehicle  silver     georgia   
3     25000       ford     door  2014  clean vehicle    blue    virginia   
4     27700  chevrolet     1500  2018  clean vehicle     red     florida   
...     ...        ...      ...   ...            ...     ...         ...   
2494   7800     nissan    versa  2019  clean vehicle     red  california   
2495   9200     nissan    versa  2018  clean vehicle  silver     florida   
2496   9200     nissan    versa  2018  clean vehicle  silver     florida   
2497   9200     nissan    versa  2018  clean vehicle   black     florida   
2498   9200     nissan    versa  2018  clean vehicle  silver     florida   

     country  acura  audi  ...  texas  utah  vermont  virginia  washington  \
0        usa      0     0  ...      0     0        0         0           0   
1        usa      0     0  ...      0     0        0         0           0   
2        usa      0     0  ...      0     0        0         0           0   
3        usa      0     0  ...      0     0        0         1           0   
4        usa      0     0  ...      0     0        0         0           0   
...      ...    ...   ...  ...    ...   ...      ...       ...         ...   
2494     usa      0     0  ...      0     0        0         0           0   
2495     usa      0     0  ...      0     0        0         0           0   
2496     usa      0     0  ...      0     0        0         0           0   
2497     usa      0     0  ...      0     0        0         0           0   
2498     usa      0     0  ...      0     0        0         0           0   

      west virginia  wisconsin  wyoming   canada   usa  
0                 0          0        0        0     1  
1                 0          0        0        0     1  
2                 0          0        0        0     1  
3                 0          0        0        0     1  
4                 0          0        0        0     1  
...             ...        ...      ...      ...   ...  
2494              0          0        0        0     1  
2495              0          0        0        0     1  
2496              0          0        0        0     1  
2497              0          0        0        0     1  
2498              0          0        0        0     1  

[2499 rows x 260 columns]

In [13]:
final = merged.drop(['brand','model','title_status','color','state','country'], axis='columns')
final

price  year  acura  audi  bmw  buick  cadillac  chevrolet  chrysler  \
0      6300  2008      0     0    0      0         0          0         0   
1      2899  2011      0     0    0      0         0          0         0   
2      5350  2018      0     0    0      0         0          0         0   
3     25000  2014      0     0    0      0         0          0         0   
4     27700  2018      0     0    0      0         0          1         0   
...     ...   ...    ...   ...  ...    ...       ...        ...       ...   
2494   7800  2019      0     0    0      0         0          0         0   
2495   9200  2018      0     0    0      0         0          0         0   
2496   9200  2018      0     0    0      0         0          0         0   
2497   9200  2018      0     0    0      0         0          0         0   
2498   9200  2018      0     0    0      0         0          0         0   

      dodge  ...  texas  utah  vermont  virginia  washington  west virginia  \
0         0  ...      0     0        0         0           0              0   
1         0  ...      0     0        0         0           0              0   
2         1  ...      0     0        0         0           0              0   
3         0  ...      0     0        0         1           0              0   
4         0  ...      0     0        0         0           0              0   
...     ...  ...    ...   ...      ...       ...         ...            ...   
2494      0  ...      0     0        0         0           0              0   
2495      0  ...      0     0        0         0           0              0   
2496      0  ...      0     0        0         0           0              0   
2497      0  ...      0     0        0         0           0              0   
2498      0  ...      0     0        0         0           0              0   

      wisconsin  wyoming   canada   usa  
0             0        0        0     1  
1             0        0        0     1  
2             0        0        0     1  
3             0        0        0     1  
4             0        0        0     1  
...         ...      ...      ...   ...  
2494          0        0        0     1  
2495          0        0        0     1  
2496          0        0        0     1  
2497          0        0        0     1  
2498          0        0        0     1  

[2499 rows x 253 columns]

In [14]:
from sklearn.linear_model import LinearRegression
model=LinearRegression()

In [17]:
X=final.drop('price', axis='columns')
X

year  acura  audi  bmw  buick  cadillac  chevrolet  chrysler  dodge  \
0     2008      0     0    0      0         0          0         0      0   
1     2011      0     0    0      0         0          0         0      0   
2     2018      0     0    0      0         0          0         0      1   
3     2014      0     0    0      0         0          0         0      0   
4     2018      0     0    0      0         0          1         0      0   
...    ...    ...   ...  ...    ...       ...        ...       ...    ...   
2494  2019      0     0    0      0         0          0         0      0   
2495  2018      0     0    0      0         0          0         0      0   
2496  2018      0     0    0      0         0          0         0      0   
2497  2018      0     0    0      0         0          0         0      0   
2498  2018      0     0    0      0         0          0         0      0   

      ford  ...  texas  utah  vermont  virginia  washington  west virginia  \
0        0  ...      0     0        0         0           0              0   
1        1  ...      0     0        0         0           0              0   
2        0  ...      0     0        0         0           0              0   
3        1  ...      0     0        0         1           0              0   
4        0  ...      0     0        0         0           0              0   
...    ...  ...    ...   ...      ...       ...         ...            ...   
2494     0  ...      0     0        0         0           0              0   
2495     0  ...      0     0        0         0           0              0   
2496     0  ...      0     0        0         0           0              0   
2497     0  ...      0     0        0         0           0              0   
2498     0  ...      0     0        0         0           0              0   

      wisconsin  wyoming   canada   usa  
0             0        0        0     1  
1             0        0        0     1  
2             0        0        0     1  
3             0        0        0     1  
4             0        0        0     1  
...         ...      ...      ...   ...  
2494          0        0        0     1  
2495          0        0        0     1  
2496          0        0        0     1  
2497          0        0        0     1  
2498          0        0        0     1  

[2499 rows x 252 columns]

In [18]:
y=final['price']
y

0        6300
1        2899
2        5350
3       25000
4       27700
        ...  
2494     7800
2495     9200
2496     9200
2497     9200
2498     9200
Name: price, Length: 2499, dtype: int64

In [19]:
model.fit(X,y)

LinearRegression()

In [20]:
model.predict(X)

array([6254.375, 2897.25 , 4568.875, ..., 5806.875, 7854.   , 5806.875])

### Now we will check how accurate our model is which can be done using ***score*** function.

In [22]:
model.score(X,y)

0.7280235575081315

This shows that the model is 72.80% accurate

## Using Sklearn ***One-Hot encoder***
The task that we did by dummy variables will be done by sklearn library, in which it will automatically assign the values to the string variables